File: apple/Sources/Kithara/KitharaPlayerItem.swift
import Combine
import Foundation
import KitharaFFI

/// A single audio item that can be queued in ``KitharaPlayer``.
///
/// All preferences (bitrate caps, ABR mode, headers) are frozen at
/// construction. Loading starts automatically when the item is inserted
/// into a ``KitharaPlayer``.
///
/// ```swift
/// let item = KitharaPlayerItem(url: "https://example.com/song.mp3")
/// try player.insert(item)
/// ```
///
/// `open` for additive subclassing (extra state, methods, protocol
/// conformances). Methods stay `public`, not `open`, so item behavior cannot
/// be overridden. Subclasses construct via the public `init(url:…)` and must
/// preserve the `@unchecked Sendable` contract: no non-`Sendable` mutable state.
open class KitharaPlayerItem: KitharaPlayerItemProtocol, @unchecked Sendable {
    /// Unique queue-item identity for SwiftUI lists.
    public nonisolated var id: Int64 { uuid }

    /// Caller-facing content track id.
    public nonisolated let audioId: TrackId

    /// Internal id allocated by the Rust queue layer.
    nonisolated let ffiTrackId: KitharaFFI.TrackId

    /// Unique queue-item handle. Distinct from ``audioId`` when the same
    /// content track is inserted more than once.
    public nonisolated var uuid: Int64 { _inner.uuidI64() }

    /// The source URL — `file://…` for absolute local paths.
    public nonisolated let url: URL

    /// Whether the source is an endless live stream.
    public nonisolated var isLiveStream: Bool { _inner.isLiveStream() }

    /// Cached duration in seconds. Defaults to `0` until the underlying
    /// resource emits a duration update.
    public nonisolated var durationSec: TimeInterval { _inner.durationSec() }

    // MARK: - Event stream

    private let _eventSubject = PassthroughSubject<ItemEvent, Never>()

    /// Single stream of all item events from Rust.
    public nonisolated var eventPublisher: AnyPublisher<ItemEvent, Never> {
        _eventSubject.eraseToAnyPublisher()
    }

    // MARK: - Discrete publishers (KitharaPlayerItemProtocol)

    /// Buffered byte ranges. Each emission is the full set. Combine
    /// equivalent of iOS `rxLoadedRanges`.
    public nonisolated var loadedRanges: AnyPublisher<[ItemLoadedRange], Never> {
        _eventSubject
            .compactMap { event -> [ItemLoadedRange]? in
                if case let .loadedRangesChanged(ranges) = event {
                    return ranges.map(ItemLoadedRange.init(ffi:))
                }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// Duration publisher (seconds). Combine equivalent of an iOS
    /// `rxDuration` (not in their current protocol but useful here).
    public nonisolated var duration: AnyPublisher<Double?, Never> {
        _eventSubject
            .compactMap { event -> Double?? in
                if case let .durationChanged(seconds) = event {
                    return .some(seconds)
                }
                return nil
            }
            .map { $0 }
            .eraseToAnyPublisher()
    }

    /// Bitrate publisher (bits/sec, latest applied variant). Combine
    /// equivalent of iOS `rxBitrate`.
    public nonisolated var bitrate: AnyPublisher<Int32, Never> {
        _eventSubject
            .compactMap { event -> Int32? in
                if case let .variantApplied(variant) = event {
                    let bps = variant.bandwidthBps
                    return Int32(clamping: bps)
                }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// HLS variant ladder discovered for this item. Fires once per
    /// master-playlist parse with the full sorted-by-bandwidth list.
    /// Use this to populate a quality picker UI.
    public nonisolated var variantsDiscovered: AnyPublisher<[Variant], Never> {
        _eventSubject
            .compactMap { event -> [Variant]? in
                if case let .variantsDiscovered(variants) = event {
                    return variants.map(Variant.init(ffi:))
                }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// Variant the ABR controller has *chosen* but not yet applied —
    /// emits before the next segment fetch on that variant.
    public nonisolated var variantSelected: AnyPublisher<Variant, Never> {
        _eventSubject
            .compactMap { event -> Variant? in
                if case let .variantSelected(variant) = event {
                    return Variant(ffi: variant)
                }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// Variant the decoder is *currently* feeding to the audio
    /// pipeline. Emits when the segment from a newly-selected variant
    /// crosses the playback head — i.e. when the user audibly hears
    /// the new quality.
    public nonisolated var variantApplied: AnyPublisher<Variant, Never> {
        _eventSubject
            .compactMap { event -> Variant? in
                if case let .variantApplied(variant) = event {
                    return Variant(ffi: variant)
                }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// Fires once the metadata layer reports `ReadyToPlay`. Combine
    /// equivalent of iOS `rxReadyToPlay`.
    public nonisolated var readyToPlay: AnyPublisher<Void, Never> {
        _eventSubject
            .compactMap { event -> Void? in
                if case let .statusChanged(status) = event, status == .readyToPlay {
                    return ()
                }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// Fires when the item plays out to its natural end. Combine
    /// equivalent of iOS `rxDidReachEnd`.
    public nonisolated var didReachEnd: AnyPublisher<Void, Never> {
        _eventSubject
            .compactMap { event -> Void? in
                if case .didReachEnd = event { return () }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// Fires when playback stalls. Combine equivalent of iOS
    /// `rxDidStall`.
    public nonisolated var didStall: AnyPublisher<Void, Never> {
        _eventSubject
            .compactMap { event -> Void? in
                if case .didStall = event { return () }
                return nil
            }
            .eraseToAnyPublisher()
    }

    /// Error publisher. Maps `ItemEvent.error` / `ItemEvent.didFail`
    /// onto ``PlayerError``. Combine equivalent of iOS `rxError`.
    public nonisolated var error: AnyPublisher<Error, Never> {
        _eventSubject
            .compactMap { event -> PlayerError? in
                switch event {
                case let .error(message): return .itemFailed(message)
                case .didFail: return .itemFailed("item did fail")
                default: return nil
                }
            }
            .map { $0 as Error }
            .eraseToAnyPublisher()
    }

    // MARK: - Bitrate preferences

    /// Preferred peak bitrate in bits per second. Zero means no limit.
    public nonisolated var preferredPeakBitrate: Double {
        _inner.preferredPeakBitrate()
    }

    /// Preferred peak bitrate for expensive networks in bits per second. Zero means no limit.
    public nonisolated var preferredPeakBitrateForExpensiveNetworks: Double {
        _inner.preferredPeakBitrateForExpensiveNetworks()
    }

    // MARK: - Internal

    nonisolated let _inner: AudioPlayerItem

    // MARK: - Init

    /// Create a new item.
    ///
    /// - Parameters:
    ///   - url: The audio source URL.
    ///   - audioId: Caller-facing content track id. If `nil`, Kithara
    ///     uses the internally allocated queue id converted to `Int`.
    ///   - uuid: Caller-facing queue item id. If `nil`, Kithara uses
    ///     its legacy generated handle.
    ///   - additionalHeaders: Optional HTTP headers included in all requests for this item.
    ///   - preferredPeakBitrate: Peak bitrate ceiling in bits/sec. `0` means no cap.
    ///   - preferredPeakBitrateForExpensiveNetworks: Peak bitrate ceiling on
    ///     expensive networks (cellular). `0` means no cap.
    ///   - abrMode: Optional per-item ABR mode override.
    public init(
        url: String,
        audioId: TrackId? = nil,
        uuid: Int64? = nil,
        additionalHeaders: [String: String]? = nil,
        preferredPeakBitrate: Double = 0,
        preferredPeakBitrateForExpensiveNetworks: Double = 0,
        abrMode: AbrMode? = nil
    ) {
        let ffiAbrMode: FfiAbrMode? = abrMode.map {
            switch $0 {
            case .auto: return .auto
            case .manual(let variantIndex): return .manual(variantIndex: UInt32(variantIndex))
            }
        }
        let config = FfiItemConfig(
            abrMode: ffiAbrMode,
            audioId: audioId.map { Self.ffiTrackId(from: $0) },
            headers: additionalHeaders,
            uuidI64: uuid,
            url: url,
            isLiveStream: false,
            preferredPeakBitrate: preferredPeakBitrate,
            preferredPeakBitrateExpensive: preferredPeakBitrateForExpensiveNetworks
        )
        self._inner = AudioPlayerItem(config: config)
        self.ffiTrackId = _inner.queueId()
        self.audioId = TrackId(ffi: _inner.audioId())
        self.url = URL(string: url) ?? URL(fileURLWithPath: url)

        let observer = ItemObserverBridge(subject: _eventSubject)
        _inner.setObserver(observer: observer)
    }

    /// Internal init wrapping an existing FFI item (used by ``KitharaPlayer/items``).
    init(inner: AudioPlayerItem) {
        self._inner = inner
        self.ffiTrackId = inner.queueId()
        self.audioId = TrackId(ffi: inner.audioId())
        let urlString = inner.url()
        self.url = URL(string: urlString) ?? URL(fileURLWithPath: urlString)

        let observer = ItemObserverBridge(subject: _eventSubject)
        inner.setObserver(observer: observer)
    }

    private static func ffiTrackId(from id: TrackId) -> KitharaFFI.TrackId {
        guard let converted = KitharaFFI.TrackId(exactly: id) else {
            preconditionFailure("Kithara TrackId \(id) cannot be represented as KitharaFFI.TrackId")
        }
        return converted
    }

    // MARK: - Load / playability

    /// Resolve a ``ItemLoadResult`` describing the item's current load
    /// status. The result reflects cached state — `KitharaPlayer.insert`
    /// already kicks off background loading.
    public func load() async -> ItemLoadResult {
        await withCheckedContinuation { continuation in
            let bridge = ItemLoadCallbackBridge { result in
                continuation.resume(returning: result)
            }
            _inner.load(callback: bridge)
        }
    }

    /// Whether the item is playable at `progress` (seconds) given the
    /// caller-supplied buffered `ranges`. Live streams are reported
    /// playable unconditionally.
    public func isPlayable(progress: Double, ranges: [ItemLoadedRange]) -> Bool {
        let ffiRanges = ranges.map { range in
            FfiTimeRange(durationSeconds: range.duration, startSeconds: range.start)
        }
        return _inner.isPlayable(progress: progress, ranges: ffiRanges)
    }

    /// `and:`-labelled overload for call sites that prefer that spelling.
    /// Delegates to ``isPlayable(progress:ranges:)``.
    public func isPlayable(progress: Double, and ranges: [ItemLoadedRange]) -> Bool {
        isPlayable(progress: progress, ranges: ranges)
    }

    /// Whether the item is playable for a structured playback progress.
    public func isPlayable(progress: PlaybackProgress, ranges: [ItemLoadedRange]) -> Bool {
        isPlayable(progress: progress, ranges: ranges, startTolerance: .zero)
    }

    /// Whether the item is playable for a structured playback progress.
    ///
    /// `startTolerance` accepts a buffered range whose start lands slightly
    /// after `progress.time`, which is useful for HLS seek readiness policies.
    public func isPlayable(
        progress: PlaybackProgress,
        ranges: [ItemLoadedRange],
        startTolerance: TimeInterval
    ) -> Bool {
        if isLiveStream {
            return true
        }
        if progress.value >= 1, !ranges.isEmpty {
            return true
        }
        let tolerance = max(TimeInterval.zero, startTolerance)
        let adjustedRanges = ranges.map { range in
            ItemLoadedRange(
                start: range.start - tolerance,
                duration: range.duration + tolerance
            )
        }
        return isPlayable(progress: progress.time, ranges: adjustedRanges)
    }
}

// MARK: - Public value types

/// Outcome reported by ``KitharaPlayerItem/load()``.
public struct ItemLoadResult: Sendable, Equatable {
    /// `true` once the metadata layer recognises encrypted segments.
    public let hasProtectedContent: Bool
    /// `true` when the item has enough metadata to start playback.
    public let isPlayable: Bool

    init(ffi: FfiItemLoadResult) {
        self.hasProtectedContent = ffi.hasProtectedContent
        self.isPlayable = ffi.isPlayable
    }
}

/// Buffered range expressed as `[start, start + duration)` seconds.
public struct ItemLoadedRange: Sendable, Equatable {
    /// Range start in seconds from the item's timeline origin.
    public let start: Double
    /// Range length in seconds.
    public let duration: Double

    /// Construct a range from its start + length in seconds.
    public init(start: Double, duration: Double) {
        self.start = start
        self.duration = duration
    }

    init(ffi: FfiTimeRange) {
        self.start = ffi.startSeconds
        self.duration = ffi.durationSeconds
    }

    public init<Source: ItemLoadedRangeSource>(_ source: Source) {
        self.start = source.start
        self.duration = source.duration
    }
}

/// Structured playback progress for queue-readiness checks.
public struct PlaybackProgress: Sendable, Equatable {
    /// Normalised playback position in `[0, 1]`.
    public let value: TimeInterval
    /// Playback position in seconds.
    public let time: TimeInterval
    /// Item duration in seconds.
    public let duration: TimeInterval

    public init(value: TimeInterval, time: TimeInterval, duration: TimeInterval) {
        self.value = value
        self.time = time
        self.duration = duration
    }

    public init<Source: PlaybackProgressSource>(_ source: Source) {
        self.value = source.value
        self.time = source.time
        self.duration = source.duration
    }
}

public protocol ItemLoadedRangeSource {
    var start: TimeInterval { get }
    var duration: TimeInterval { get }
}

public protocol PlaybackProgressSource {
    var value: TimeInterval { get }
    var time: TimeInterval { get }
    var duration: TimeInterval { get }
}

extension ItemLoadedRange: ItemLoadedRangeSource {}
extension PlaybackProgress: PlaybackProgressSource {}

// MARK: - ItemObserver bridge (single on_event callback)

private final class ItemObserverBridge: KitharaFFI.ItemObserver, @unchecked Sendable {
    private let subject: PassthroughSubject<ItemEvent, Never>

    init(subject: PassthroughSubject<ItemEvent, Never>) {
        self.subject = subject
    }

    func onEvent(event: FfiItemEvent) {
        subject.send(event)
    }
}

// MARK: - ItemLoadCallback bridge

private final class ItemLoadCallbackBridge: KitharaFFI.ItemLoadCallback, @unchecked Sendable {
    private let handler: (ItemLoadResult) -> Void

    init(handler: @escaping (ItemLoadResult) -> Void) {
        self.handler = handler
    }

    func onComplete(result: FfiItemLoadResult) {
        handler(ItemLoadResult(ffi: result))
    }
}

File: android/lib/src/main/kotlin/com/kithara/KitharaPlayerItem.kt
package com.kithara

import com.kithara.ffi.AudioPlayerItem as FfiAudioPlayerItem
import com.kithara.ffi.FfiAbrMode
import com.kithara.ffi.FfiItemConfig
import com.kithara.ffi.FfiItemEvent
import com.kithara.ffi.FfiItemLoadResult
import com.kithara.ffi.FfiItemStatus
import com.kithara.ffi.FfiSourceSettings
import com.kithara.ffi.FfiTimeRange
import com.kithara.ffi.FfiVariant
import com.kithara.ffi.ItemLoadCallback
import com.kithara.ffi.ItemObserver
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.flow.onStart

private fun itemConfig(
    url: String,
    additionalHeaders: Map<String, String>?,
    preferredPeakBitrate: Double,
    preferredPeakBitrateForExpensiveNetworks: Double,
    abrMode: FfiAbrMode?,
    isLiveStream: Boolean,
    audioId: ULong?,
    uuid: Long?,
): FfiItemConfig = FfiItemConfig(
    abrMode = abrMode,
    audioId = audioId,
    headers = additionalHeaders,
    uuidI64 = uuid,
    url = url,
    preferredPeakBitrate = preferredPeakBitrate,
    preferredPeakBitrateExpensive = preferredPeakBitrateForExpensiveNetworks,
    isLiveStream = isLiveStream,
)

/**
 * A single audio item that can be queued in [KitharaPlayer].
 *
 * All preferences (bitrate caps, ABR mode, headers) are frozen at
 * construction. Loading starts automatically when the item is inserted
 * into a [KitharaPlayer].
 *
 * ```kotlin
 * val item = KitharaPlayerItem("https://example.com/audio.mp3")
 * player.insert(item)
 * player.play()
 * ```
 */
class KitharaPlayerItem internal constructor(
    internal val inner: FfiAudioPlayerItem,
) {
    constructor(
        url: String,
        additionalHeaders: Map<String, String>? = null,
        preferredPeakBitrate: Double = 0.0,
        preferredPeakBitrateForExpensiveNetworks: Double = 0.0,
        abrMode: FfiAbrMode? = null,
        isLiveStream: Boolean = false,
        audioId: ULong? = null,
        uuid: Long? = null,
    ) : this(
        FfiAudioPlayerItem(
            itemConfig(
                abrMode = abrMode,
                audioId = audioId,
                additionalHeaders = additionalHeaders,
                uuid = uuid,
                url = url,
                preferredPeakBitrate = preferredPeakBitrate,
                preferredPeakBitrateForExpensiveNetworks = preferredPeakBitrateForExpensiveNetworks,
                isLiveStream = isLiveStream,
            )
        )
    )

    /** Create an item with generated source settings validated before loading. */
    constructor(
        url: String,
        additionalHeaders: Map<String, String>? = null,
        preferredPeakBitrate: Double = 0.0,
        preferredPeakBitrateForExpensiveNetworks: Double = 0.0,
        abrMode: FfiAbrMode? = null,
        isLiveStream: Boolean = false,
        audioId: ULong? = null,
        uuid: Long? = null,
        sourceSettings: FfiSourceSettings,
    ) : this(
        FfiAudioPlayerItem.newWithSourceSettings(
            config = itemConfig(
                abrMode = abrMode,
                audioId = audioId,
                additionalHeaders = additionalHeaders,
                uuid = uuid,
                url = url,
                preferredPeakBitrate = preferredPeakBitrate,
                preferredPeakBitrateForExpensiveNetworks = preferredPeakBitrateForExpensiveNetworks,
                isLiveStream = isLiveStream,
            ),
            settings = sourceSettings,
        )
    )

    /** Source URL for this item. */
    val url: String = inner.url()

    /**
     * Stable per-item identifier: the decimal string form of the
     * monotonic `TrackId` (`u64`) the queue reserves at construction.
     * Mirrors the iOS `AudioPlayerItemProtocol.audioId`. Synonym alias
     * [id] is kept for `Identifiable`-style consumers.
     */
    val audioId: String = inner.audioId().toString()

    /** Synonym for [audioId]. */
    val id: String get() = audioId

    private val queueId: ULong = inner.queueId()

    /** Two wrappers are equal when they stand for the same queued item. */
    override fun equals(other: Any?): Boolean =
        other is KitharaPlayerItem && other.queueId == queueId

    override fun hashCode(): Int = queueId.hashCode()

    /** Numeric form of [audioId] derived from the first 16 hex digits. */
    val uuid: Long
        get() = inner.uuidI64()

    /**
     * Caller-declared live-stream flag. Mirrors the iOS
     * `AudioPlayerItemProtocol.isLiveStream`.
     */
    val isLiveStream: Boolean
        get() = inner.isLiveStream()

    /**
     * Cached duration in seconds. Defaults to `0.0` until the
     * underlying resource emits a duration update.
     */
    val durationSec: Double
        get() = inner.durationSec()

    /**
     * Item state, emitted on every change while collected. The first
     * value is the state as it stands when collection starts.
     */
    val state: Flow<ItemState>
        get() = ffiEvents
            .filter { it.changesState() }
            .map { snapshot() }
            .onStart { emit(snapshot()) }
            .distinctUntilChanged()

    /** One-shot item events. */
    val events: Flow<KitharaItemEvent>
        get() = ffiEvents.mapNotNull { it.toKitharaItemEvent() }

    val status: ItemStatus
        get() = snapshot().status

    val duration: Double?
        get() = snapshot().duration

    /** Buffered ranges (start + duration in seconds). */
    val loadedRanges: List<ItemLoadedRange>
        get() = snapshot().loadedRanges

    val error: KitharaError?
        get() = snapshot().error

    /** Preferred peak bitrate in bits per second. Zero means no limit. */
    val preferredPeakBitrate: Double
        get() = inner.preferredPeakBitrate()

    /** Preferred peak bitrate on expensive networks. Zero means no limit. */
    val preferredPeakBitrateForExpensiveNetworks: Double
        get() = inner.preferredPeakBitrateForExpensiveNetworks()

    /**
     * Resolve a snapshot of the current load status. The result
     * reflects cached state — `KitharaPlayer.insert` already starts
     * background loading. Mirrors the iOS `func load() -> Observable<…>`
     * via Kotlin coroutines.
     */
    suspend fun load(): ItemLoadResult = suspendCoroutine { cont ->
        inner.load(object : ItemLoadCallback {
            override fun onComplete(result: FfiItemLoadResult) {
                cont.resume(
                    ItemLoadResult(
                        hasProtectedContent = result.hasProtectedContent,
                        isPlayable = result.isPlayable,
                    )
                )
            }
        })
    }

    /**
     * Whether the item is playable at `progress` (seconds) given the
     * caller-supplied buffered `ranges`. Live streams are reported
     * playable unconditionally.
     */
    fun isPlayable(progress: Double, ranges: List<ItemLoadedRange>): Boolean =
        inner.isPlayable(
            progress,
            ranges.map { FfiTimeRange(durationSeconds = it.duration, startSeconds = it.start) },
        )

    /**
     * Cold: each collector registers its own observer on the item. The
     * native event thread never waits on a collector; a slow collector
     * loses the oldest buffered events.
     */
    private val ffiEvents: Flow<FfiItemEvent> = callbackFlow {
        val observer = object : ItemObserver {
            override fun onEvent(event: FfiItemEvent) {
                trySend(event)
            }
        }
        val id = inner.addObserver(observer)
        awaitClose { inner.removeObserver(id) }
    }.buffer(EVENT_BUFFER, BufferOverflow.DROP_OLDEST)

    private fun FfiItemEvent.toKitharaItemEvent(): KitharaItemEvent? = when (this) {
        is FfiItemEvent.DurationChanged -> KitharaItemEvent.DurationChanged(seconds)
        is FfiItemEvent.VariantsDiscovered -> KitharaItemEvent.VariantsDiscovered(variants.map { it.toKitharaVariant() })
        is FfiItemEvent.VariantSelected -> KitharaItemEvent.VariantSelected(variant.toKitharaVariant())
        is FfiItemEvent.VariantApplied -> KitharaItemEvent.VariantApplied(variant.toKitharaVariant())
        is FfiItemEvent.Error -> KitharaItemEvent.Error(error)
        else -> null
    }

    private companion object {
        const val EVENT_BUFFER = 64
    }

    private fun snapshot(): ItemState {
        val state = inner.state()
        return ItemState(
            loadedRanges = state.loadedRanges.map {
                ItemLoadedRange(start = it.startSeconds, duration = it.durationSeconds)
            },
            duration = state.durationSeconds,
            error = state.error?.let { KitharaError.ItemFailed(it) },
            status = state.status.toItemStatus(),
        )
    }
}

/** Outcome reported by [KitharaPlayerItem.load]. */
data class ItemLoadResult(
    val hasProtectedContent: Boolean,
    val isPlayable: Boolean,
)

/** Buffered range expressed as `[start, start + duration)` seconds. */
data class ItemLoadedRange(
    val start: Double,
    val duration: Double,
)

private fun FfiItemEvent.changesState(): Boolean = when (this) {
    is FfiItemEvent.StatusChanged,
    is FfiItemEvent.DurationChanged,
    is FfiItemEvent.LoadedRangesChanged,
    is FfiItemEvent.DidFail,
    is FfiItemEvent.Error -> true
    else -> false
}

private fun FfiVariant.toKitharaVariant(): KitharaVariant =
    KitharaVariant(index, bandwidthBps.toLong(), name)

private fun FfiItemStatus.toItemStatus(): ItemStatus = when (this) {
    FfiItemStatus.READY_TO_PLAY -> ItemStatus.ReadyToPlay
    FfiItemStatus.FAILED -> ItemStatus.Failed
    FfiItemStatus.UNKNOWN -> ItemStatus.Unknown
}
